### PR TITLE
[Parse] Statement typo corrections

### PR DESCRIPTION
**Pull request by kitaisreal**
Current implementation steps:
1. If consecutive statement error occurred we are getting previous identifier, then trying to typo correct it to one of the statements keywords.
2. Updated tests

Current implementation is limited to only identifiers that start from the beginning of the line. Like
``` swift
gaurd x>5 else { return }
```
The limitation exists because it is ambigous how to handle case where 
``` swift
x=1+i i=i+1
```
Should we correct it to `if` statement, or provide insert of `';'`.

https://bugs.swift.org/browse/SR-8510

### CODE REVIEW

File: lib/Parse/ParseStmt.cpp [hunk 1]
```diff
@@ -2529,3 +2548,66 @@ ParserResult<Stmt> Parser::parseStmtPoundAssert() {
...
+/////
+///// \param identifier The name of the identifier.
+///// \param stmt The name of the argument.
+///// \param maxScore The maximum score permitted by this comparison, or
+///// 0 if there is no limit.
+/////
+///// \returns the score, if it is good enough to even consider this a match.
+///// Otherwise, an empty optional.
+/////
+static Optional<unsigned> scoreIdentifierAndStmtNameTypo(StringRef identifier,
```
Comment (kitaisreal) [inline lib/Parse/ParseStmt.cpp]: There are identical function used for matching stringRefs in CSSimplify.cpp. Can it be moved to StringExtras?

File: include/swift/AST/DiagnosticsParse.def [hunk 2]
```diff
@@ -925,6 +925,8 @@ ERROR(statement_begins_with_closure,none,
       "top-level statement cannot begin with a closure expression", ())
 ERROR(statement_same_line_without_semi,none,
       "consecutive statements on a line must be separated by ';'", ())
+ERROR(statement_misspell,none,
+    "did you misspell '%0'", (StringRef))
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsParse.def]: Some nits:
1) The phrasing here is of a note, but the diagnostic is an error.  Take a look at how we do typo correction diagnostics in libSema.
2) Rhetorical questions need punctuation

File: test/Parse/statement_typo.swift [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// RUN: %target-typecheck-verify-swift
+
+func statementMisspelledAtTheBegginingOfTheLine() {
+  let test = 5
+  gaurd test > 5 else { return } // expected-error {{did you misspell 'guard'}} {{3-8=guard}} expected-error {{use of unresolved identifier 'gaurd'}}
+  i test > 5 { return } // expected-error {{did you misspell 'if'}} {{3-4=if}} expected-error {{use of unresolved identifier 'i'}}
+  whle test > 5 { return } // expected-error {{did you misspell 'while'}} {{3-7=while}} expected-error {{use of unresolved identifier 'whle'}}
```
Comment (CodaFi) [inline test/Parse/statement_typo.swift]: Can we recover better here?  (Would it make sense to use the nearest-neighbor match to start parsing the corresponding statement kind?)

File: lib/Parse/ParseStmt.cpp (hunk 1 above)
Comment (CodaFi) [reply lib/Parse/ParseStmt.cpp]: I think you can get away with a table of keywords and a loop checking `edit_distance`.

File: lib/Parse/ParseStmt.cpp [hunk 4]
```diff
@@ -364,10 +364,29 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
...
+      if (previousToken.getKind() == tok::identifier) {
+        SourceLoc startOfLineLocation = L->getLocForStartOfLine(SourceMgr, previousToken.getLoc());
+        SourceLoc locationOfTokenAtStartOfLine = Lexer::getTokenAtLocation(SourceMgr, startOfLineLocation).getLoc();
+        if (previousToken.getLoc() == locationOfTokenAtStartOfLine) {
+          stmtTokMatch = closestMatchingStmtTokenForIdentifier(previousToken.getText());
+        }
+      }
+      if (stmtTokMatch) {
+        ParserPosition parserPosition = ParserPosition(L->getStateForBeginningOfToken(previousToken), PreviousLoc);
+        backtrackToPosition(parserPosition);
```
Comment (CodaFi) [inline lib/Parse/ParseStmt.cpp]: I don’t think you have to backtrack.  You only need the location of the typo’d keyword.

**Comment by CodaFi** [comment]
I left some high-level comments about the approach here.  I think @rintaro is more qualified than I am to get into the finer details.

Thanks for tackling this.  It’s a tough nut to crack.

File: test/Parse/statement_typo.swift [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// RUN: %target-typecheck-verify-swift
+
+func statementMisspelledAtTheBegginingOfTheLine() {
```
Comment (CodaFi) [inline test/Parse/statement_typo.swift]: Can you add the functions from the SR too?

File: test/Parse/statement_typo.swift [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  let test = 5
+  gaurd test > 5 else { return } // expected-error {{did you misspell 'guard'}} {{3-8=guard}} expected-error {{use of unresolved identifier 'gaurd'}}
+  i test > 5 { return } // expected-error {{did you misspell 'if'}} {{3-4=if}} expected-error {{use of unresolved identifier 'i'}}
+  whle test > 5 { return } // expected-error {{did you misspell 'while'}} {{3-7=while}} expected-error {{use of unresolved identifier 'whle'}}
+}
+
+func statementMisspelledNotAtTheBegginingOfTheLine() {
+  let i = 5
+  var x = i x=x+1 // expected-error {{consecutive statements}}
+}
```
Comment (CodaFi) [inline test/Parse/statement_typo.swift]: Nit: Trailing newline please.

File: test/Parse/statement_typo.swift (hunk 5 above)
Comment (kitaisreal) [reply test/Parse/statement_typo.swift]: Sure. Added.

File: test/Parse/statement_typo.swift (hunk 6 above)
Comment (kitaisreal) [reply test/Parse/statement_typo.swift]: Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 1 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Agree. Fixed.

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Agree. Fixed.

File: test/Parse/statement_typo.swift (hunk 3 above)
Comment (kitaisreal) [reply test/Parse/statement_typo.swift]: Yes. Fixed. Please check my comment below.

File: lib/Parse/ParseStmt.cpp (hunk 4 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Currently I used backtrack to update previous token with our corrected token kind and continue parsing as there were no error. 
Maybe there are better way to recover there for example just skip tokens until we met end of statement.
Do we have any other options @rintaro ?

File: lib/Parse/ParseStmt.cpp [hunk 7]
```diff
@@ -368,6 +368,25 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
...
+      Optional<tok> stmtTokMatch = None;
+      if (previousToken.getKind() == tok::identifier) {
+        SourceLoc startOfLineLocation = L->getLocForStartOfLine(SourceMgr, previousToken.getLoc());
+        SourceLoc locationOfTokenAtStartOfLine = Lexer::getTokenAtLocation(SourceMgr, startOfLineLocation).getLoc();
+        if (previousToken.getLoc() == locationOfTokenAtStartOfLine) {
+          stmtTokMatch = closestMatchingStmtTokenForIdentifier(previousToken.getText());
+        }
+      }
+      if (stmtTokMatch) {
+        Entries.pop_back();
```
Comment (kitaisreal) [inline lib/Parse/ParseStmt.cpp]: We are removing latest entry, because it is `unresolved identifier` expression from parse expression. If we will not remove it `undefined identifier` error will be emitted by Sema.

File: include/swift/AST/DiagnosticsParse.def (hunk 2 above)
Comment (kitaisreal) [reply include/swift/AST/DiagnosticsParse.def]: Agree. Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 4 above)
Comment (rintaro) [reply lib/Parse/ParseStmt.cpp]: Unfortunately,`setKind()` to a keyword kind is not not an option because we have an invariant in libSyntax where keyword tokens have always have the fixed text.

An option (which @CodaFi is suggesting)  is to make `parseStmt***` functions receive keyword location like
```cpp
ParserResult<Stmt> Parser::parseStmtReturn(SourceLoc tryLoc, SourceLoc ReturnLoc) {
  if (ReturnLoc.isInvalid())
    ReturnLoc = consumeToken(tok::kw_return);

  ...
}
```
Then directly call them from this recovery logic. But that's probably not simple changes considering `TopLevelCodeDecl` handling.

So, I recommend to skip to the next statement, decl, or '}' using `skipUntilDeclStmtRBrace()`. Downside of this is that we loose the diagnostics and semantic syntax highlighting until the next statement. But I think that's better than lousy diagnostics.

File: lib/Parse/ParseStmt.cpp [hunk 8]
```diff
@@ -2529,3 +2548,28 @@ ParserResult<Stmt> Parser::parseStmtPoundAssert() {
   return makeParserResult<Stmt>(new (Context) PoundAssertStmt(
       SourceRange(startLoc, endLoc), conditionExprResult.get(), message));
 }
+
+Optional<tok> Parser::closestMatchingStmtTokenForIdentifier(StringRef identifier) {
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Do we need to make it a Parser method? If not, please make this a `static` function.

File: lib/Parse/ParseStmt.cpp (hunk 4 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Hi, totally agree. I also thought that approach with backtracking and setting kind is too unnatural. Related to skipUntilDeclStmtRBrace(). If we start using such diagnostics there are case like this.

Comment (rintaro) [reply lib/Parse/ParseStmt.cpp]: I think the implementation of `skipSingle()` is *not* wrong. This is for example:
```swift
func foo() {
  if /*HERE*/ foo {
    return 1
  }
  return 2
}
```
When skipping from `/*HERE*/` to the next statement, we don't care about the contents of the `if` statement. We just want to skip to `return 1` at line 5.

> Introduce function possibleStartOfStmt() and update skipUntilDeclStmtRBrace.

Could you elaborate on this? How will `possibleStartOfStmt()` be different from `isStartOfStmt()`?

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Hi, @rintaro, in case of
``` swift
func bar() {
    let i = 5
    /*HERE*/ gaurd i > 5 else { return }
    gaurd i > 5 else { return }
}
```
We are starting from /*HERE*/ we find out that there are typo error and then we try to skip till the next statement.
We are skipping brackets and then we meet second guard but it is not a start of statement (because it is wrong typed), so we are just skipping until we met correct start of statement (like guard, while without error in typing).
My question is should we consider such case, when there are more than 1 statement in a row typed wrong ? 
If yes, then possibleStartOfStmt() will also try to typo statement, if it is for example gaurd we are matching guard and not skipping it.

Comment (rintaro) [reply lib/Parse/ParseStmt.cpp]: I don't think it's worth it. For now, let's use `skipUntilDeclStmtRBrace()` as is. In your test cases, just guard them by some other decl or statement like:
```swift
    gaurd i > 5 else { return }
    func parserSync() {}
    i i > 5 else { return }
```
or put them in dedicated braces like:
```swift
    do {
      gaurd i > 5 else { return }
    }
    do {
      i i > 5 else { return }
    }
```

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Ok, agree. Lets keep it simple.

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Hi, @rintaro after investigation I find out there simple usage of skipUntilDeclStmtRBrace() is not sufficient for cases like
```swift
// SR-8510
func foo() -> Int? { return 0 }
func bar() {
  gaurd /*HERE*/ let x = foo() else { return }  // expected-error {{consecutive statements}} // expected-note {{did you misspell 'guard'}} {{3-8=guard}}
  print(x)
}
```
We will stop here and then emit a bunch of unrelated errors. Are there any workarounds around this, if we are trying to use skipUntilDeclStmtRBrace() ?
Should we consider introducing another skipping function ?
```swift
void Parser::skipUntilStmtRBrace() {
  while (Tok.isNot(tok::eof, tok::r_brace, tok::pound_endif,
                 tok::pound_else, tok::pound_elseif,
                 tok::code_complete) &&
       !isStartOfStmt())
  skipSingle();
}
```

File: lib/Parse/ParseStmt.cpp [hunk 9]
```diff
@@ -368,6 +368,25 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
       diagnose(EndOfPreviousLoc, diag::statement_same_line_without_semi)
         .fixItInsert(EndOfPreviousLoc, ";");
       // FIXME: Add semicolon to the AST?
+      Token previousToken = Lexer::getTokenAtLocation(SourceMgr, PreviousLoc);
+      Optional<tok> stmtTokMatch = None;
+      if (previousToken.getKind() == tok::identifier) {
+        SourceLoc startOfLineLocation = L->getLocForStartOfLine(SourceMgr, previousToken.getLoc());
```
Comment (kitaisreal) [inline lib/Parse/ParseStmt.cpp]: Also @rintaro if we are trying to use simple skipping solution, should we remove this check for start of the line. Example:

func foo() -> Int? { return 0 }
func bar() {
  let i = 5; gaurd let x = foo() else { return }  // expected-error {{consecutive statements}} // expected-note {{did you misspell 'guard'}} {{3-8=guard}}
  print(x)
}
Should we consider it ?

File: lib/Parse/ParseStmt.cpp (hunk 8 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Sure, fixed

File: lib/Parse/ParseStmt.cpp (hunk 9 above)
Comment (rintaro) [reply lib/Parse/ParseStmt.cpp]: For now, how about just skipping the first `Tok.isAny(tok::kw_let, tok::kw_var)` in this recovery logic? like:
```
if (Tok.isAny(tok::kw_let, tok::kw_var))
  consumeToken();
skipUntilDeclStmtRBrace()
```
It's not great for something like:
```swift
gurad
  let a = someOpt1,
  let b = someOpt2
else {
}
```
But I think supporting single `let`/`var` is good enough because this is just a recovery.

File: lib/Parse/ParseStmt.cpp (hunk 4 above)
Comment (rintaro) [reply lib/Parse/ParseStmt.cpp]: For now, how about just skipping the first `Tok.isAny(tok::kw_let, tok::kw_var)` in this recovery logic? like:
```
if (Tok.isAny(tok::kw_let, tok::kw_var))
  consumeToken();
skipUntilDeclStmtRBrace()
```
It's not great for something like:
```swift
gurad
  let a = someOpt1,
  let b = someOpt2
else {
}
```
But I think supporting single `let`/`var` is good enough because this is just a recovery.

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Hi, @rintaro updated pull request. Currently recovery logic will work only if next token is let/var and wrong typed statement is on the beginning of the line. Also updated tests.

File: lib/Parse/ParseStmt.cpp (hunk 9 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: @rintaro Sure. Updated.

File: lib/Parse/ParseStmt.cpp [hunk 10]
```diff
@@ -348,6 +374,23 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
...
+        if (previousToken.getLoc() == locationOfTokenAtStartOfLine) {
+          stmtTokMatch = closestMatchingStmtTokenForIdentifier(previousToken.getText());
+        }
+      }
+      if (stmtTokMatch && Tok.isAny(tok::kw_let, tok::kw_var)) {
+        StringRef stmtTokenText = getTokenText(*stmtTokMatch);
+        diagnose(PreviousLoc, diag::statement_misspell, stmtTokenText)
+          .fixItReplace(PreviousLoc, stmtTokenText);
+        consumeToken();
+        skipUntilDeclStmtRBrace(tok::r_brace);
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Ah, I meant something like:
```cpp
if (stmtTokMatch) {
  ... diagnose ...
  if (Tok.isAny(tok::kw_let, tok::kw_var))
    consumeToken();
  skipUntilDeclStmtRBrace(tok::r_brance);
}
```
We still can recover non `let`/`var` binding cases.

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: @rintaro Sure. Fixed.

**Comment by rintaro** [comment]
Sorry for the delay. Could you add back test cases you deleted in previous commit?
```swift
func statementMisspelledAtTheBegginingOfTheLine() {
   let test = 5
   gaurd test > 5 else { return } // expected-error {{consecutive statements}} // expected-note {{did you misspell 'guard'}} {{3-8=guard}}
   i test > 5 { return } // expected-error {{consecutive statements}} // expected-note {{did you misspell 'if'}} {{3-4=if}}
   whle test > 5 { return } // expected-error {{consecutive statements}} // expected-note {{did you misspell 'while'}} {{3-7=while}}
 }

 func statementMisspelledNotAtTheBegginingOfTheLine() {
   let i = 5
   var x = i x=x+1 // expected-error {{consecutive statements}}
 }
```
These should work now.

File: include/swift/Parse/Parser.h [hunk 11]
```diff
@@ -870,7 +870,7 @@ class Parser {
                            IsFollowingGuard);
   }
   ParserResult<BraceStmt> parseBraceItemList(Diag<> ID);
-  
+
```
Comment (rintaro) [inline include/swift/Parse/Parser.h]: Unrelated change?

File: test/Parse/statement_typo.swift [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  // expected-note@-1 {{did you misspell 'guard'}} {{3-8=guard}}
+  // expected-error@-2 {{use of unresolved identifier 'gaurd'}}
+  print(x)
+}
+
+func stmtMisspelledAtTheBegginingOfTheLineWithLet() {
+  gaurd let firstWithLet = foo() else { return } // expected-error {{consecutive statements}}
+  // expected-note@-1 {{did you misspell 'guard'}} {{3-8=guard}}
+  // expected-error@-2 {{use of unresolved identifier 'gaurd'}}
+  i let secondWithLet = foo() { return } // expected-error {{consecutive statements}}
```
Comment (rintaro) [inline test/Parse/statement_typo.swift]: I suspect the previous recovery skips `i` here. Because, `use of unresolved identifier 'I'` is not emitted.
Could you split test cases to their own functions? like:
```
func stmtMisspelledAtTheBegginingOfTheLineWithLet1() {
   gaurd let firstWithLet = foo() else { return }
}
func stmtMisspelledAtTheBegginingOfTheLineWithLet2() {
   i let secondWithLet = foo() { return }
}
```

File: lib/Parse/ParseStmt.cpp [hunk 13]
```diff
@@ -194,6 +194,32 @@ static bool isAtStartOfSwitchCase(Parser &parser,
...
+  SmallVector<std::pair<StringRef, tok>, 19> stmtKeywords;
+  #define STMT_KEYWORD(kw) stmtKeywords.push_back({ #kw, tok::kw_##kw });
+  #include "swift/Syntax/TokenKinds.def"
+  unsigned bestScore = 0;
+  Optional<tok> best = None;
+  for (size_t i = 0; i < stmtKeywords.size(); ++i) {
+    std::pair<StringRef, tok> pair = stmtKeywords[i];
+    StringRef potentialStatement = pair.first;
+
+    unsigned dist = potentialStatement.edit_distance(identifier, /*AllowReplacements=*/true, /*MaxEditDistance=*/bestScore);
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns

File: lib/Parse/ParseStmt.cpp [hunk 14]
```diff
@@ -348,6 +374,25 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
       diagnose(EndOfPreviousLoc, diag::statement_same_line_without_semi)
         .fixItInsert(EndOfPreviousLoc, ";");
       // FIXME: Add semicolon to the AST?
+      Token previousToken = Lexer::getTokenAtLocation(SourceMgr, PreviousLoc);
+      Optional<tok> stmtTokMatch = None;
+      if (previousToken.getKind() == tok::identifier) {
+        SourceLoc startOfLineLocation = L->getLocForStartOfLine(SourceMgr, previousToken.getLoc());
+        SourceLoc locationOfTokenAtStartOfLine = Lexer::getTokenAtLocation(SourceMgr, startOfLineLocation).getLoc();
+        if (previousToken.getLoc() == locationOfTokenAtStartOfLine) {
+          stmtTokMatch = closestMatchingStmtTokenForIdentifier(previousToken.getText());
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns

File: lib/Parse/ParseStmt.cpp [hunk 15]
```diff
@@ -194,6 +194,32 @@ static bool isAtStartOfSwitchCase(Parser &parser,
   return parser.Tok.isAny(tok::kw_case, tok::kw_default);
 }
 
+
+static Optional<tok> closestMatchingStmtTokenForIdentifier(StringRef identifier) {
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns

**Review by rintaro — COMMENTED** [review]
Implementation looks good!
For code-formatting, I recommend [`clang-format`](https://clang.llvm.org/docs/ClangFormat.html) with `git-clang-format`.

File: lib/Parse/ParseStmt.cpp (hunk 15 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 13 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 14 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

File: include/swift/Parse/Parser.h (hunk 11 above)
Comment (kitaisreal) [reply include/swift/Parse/Parser.h]: It seems like there was trailing white space and xCode automatically trimmed it.

File: test/Parse/statement_typo.swift (hunk 12 above)
Comment (kitaisreal) [reply test/Parse/statement_typo.swift]: Sure. Fixed.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by kitaisreal** [comment]
Hi, @rintaro, what is status of this ticket ? Should be consider some different implementation ?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
